### PR TITLE
proxmox: create a common base

### PR DESCRIPTION
**Pull request by Aversiste**
##### SUMMARY

Add a doc_fragment to share the documentation regarding authentication
parameters (`api_host`, `api_user`, `api_password`, `api_token_id`,
`api_token_secret` as well as the lone `validate_certs`).

Add a module_utils to hold common code such as the argument spec (again
related to authentication parameters), a helper function to convert from
Proxmox boolean representation to python and the base class
`ProxmoxAnsible`.
For now it only handles the connection to Proxmox VE API but more can be
added in the future.

To check if everything is well in place add three new modules:
`proxmox_domain_info`, `proxmox_group_info` and `proxmox_user_info`.

And finally tests these new modules.

##### ISSUE TYPE

- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME

- proxmox_domain_info
- proxmox_group_info
- proxmox_user_info

### CODE REVIEW

**Comment by Aversiste** [comment]
With this PR I would like to exercise at best the shared elements like the doc_fragments and module_utils pieces before incorporating them in `proxmox_kvm` or `proxmox_template`.

File: changelogs/fragments/1331-proxmox-info-modules.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+minor_changes:
+  - proxmox - create a shared doc_fragments (https://github.com/ansible-collections/community.general/pull/1331)
+  - proxmox - create a shared module_utils (https://github.com/ansible-collections/community.general/pull/1331)
+  - proxmox_domain_info - new module (https://github.com/ansible-collections/community.general/pull/1331)
```
Comment (felixfontein) [inline changelogs/fragments/1331-proxmox-info-modules.yml:5]: New modules don't need changelog fragments.

File: changelogs/fragments/1331-proxmox-info-modules.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+minor_changes:
+  - proxmox - create a shared doc_fragments (https://github.com/ansible-collections/community.general/pull/1331)
+  - proxmox - create a shared module_utils (https://github.com/ansible-collections/community.general/pull/1331)
```
Comment (felixfontein) [inline changelogs/fragments/1331-proxmox-info-modules.yml:4]:
```suggestion
  - proxmox - extract common code and documentation (https://github.com/ansible-collections/community.general/pull/1331).
```
Users don't need to know what module_utils and doc_fragments are, and experts can look at the provided link :)

File: tests/integration/targets/proxmox/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,111 @@
+####################################################################
+# WARNING: These are designed specifically for Ansible tests       #
+# and should not be used as examples of how to write Ansible roles #
+####################################################################
+
+# Copyright: (c) 2020, Tristan Le Guern <tleguern at bouledef.eu>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: List domains
+  community.general.proxmox_domain_info:
```
Comment (felixfontein) [inline tests/integration/targets/proxmox/tasks/main.yml:10]: FQCN isn't needed here (thanks to some ansible-test magic). These are only needed in tests for plugins that are no action plugins or modules.

File: changelogs/fragments/1331-proxmox-info-modules.yml (hunk 1 above)
Comment (Aversiste) [reply changelogs/fragments/1331-proxmox-info-modules.yml]: Fixed.

File: tests/integration/targets/proxmox/tasks/main.yml (hunk 3 above)
Comment (Aversiste) [reply tests/integration/targets/proxmox/tasks/main.yml]: Fixed. I did not even noticed with the tasks further down the file.

**Comment by Aversiste** [comment]
I also want to give credit where credit is due to @JuddTracy-DAS whose work on [proxmox_kvm_info](https://github.com/JuddTracy-DAS/community.general/commit/ca8162c9495237e9757bfa00b5b36c3f59fe3ce5) largely started this PR.

**Comment by JuddTracy-DAS** [comment]
@Aversiste thanks.  I looked at your code real quick and it looks good.  I will try it when I get some time.

BTW did you ever send those patches for the proxmox_kvm_info module documentation or is that waiting on the shared part of this PR to be implemented?

**Comment by Aversiste** [comment]
> @Aversiste thanks. I looked at your code real quick and it looks good. I will try it when I get some time.
> 
> BTW did you ever send those patches for the proxmox_kvm_info module documentation or is that waiting on the shared part of this PR to be implemented?

I opened a PR on your fork a few days ago ! It took me a bit more time than originally announced -> https://github.com/JuddTracy-DAS/community.general/pull/2

File: plugins/modules/cloud/misc/proxmox_domain_info.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: proxmox_domain_info
+short_description: Retrieve information about one or more Proxmox VE domains
+description:
+  - Retrieve information about one or more Proxmox VE domains
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_domain_info.py:16]:
```suggestion
  - Retrieve information about one or more Proxmox VE domains.
```

File: plugins/modules/cloud/misc/proxmox_domain_info.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    aliases: ['realm', 'name']
+    type: str
+author: Tristan Le Guern (@Aversiste)
+extends_documentation_fragment: community.general.proxmox.documentation
+'''
+
+
+EXAMPLES = '''
+- name: List existing domains
+  proxmox_domain_info:
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_domain_info.py:30]: All examples need to use FQCN:
```suggestion
  community.general.proxmox_domain_info:
```

**Comment by felixfontein** [comment]
@Aversiste @JuddTracy-DAS since you're actively working on the proxmox modules and plugins, and since you're attempting to add several new modules: what do you think of moving the proxmox content to its own collection, maybe community.proxmox? This would also mean that you have your own repo, where collaboration would be a lot easier. (In particular you could have commit rights and just merge stuff without having to wait for @Andersson007 or me to look at stuff.) What do you think?

We're currently moving several things out, see https://github.com/ansible-collections/overview/discussions/117#discussioncomment-108048

**Comment by JuddTracy-DAS** [comment]
@felixfontein I am not against it, but I haven't had a lot of time for proxmox module development lately.  It looks like @Aversiste is the only one doing any real development on proxmox modules lately.  I could certainly help out but I don't know if we have enough resources to handle the full life cycle of a collection.

**Comment by Aversiste** [comment]
@felixfontein I will need some time to think about it if it is ok with you.

**Comment by felixfontein** [comment]
@JuddTracy-DAS @Aversiste I (or someone from the Ansible Community Team) would take care of the migration and setting up CI (GitHub Action based), and I can also handle releases for some time (they're pretty easy), so for you it wouldn't mean much a difference to now from the amount of work.

But take your time, think about this, there's no need to hurry :) (Well, except that it should be done maybe by the end of this year, so the plugin can be removed from community.general 2.0.0 with some advance warning :) )

File: plugins/modules/cloud/misc/proxmox_domain_info.py (hunk 5 above)
Comment (Aversiste) [reply plugins/modules/cloud/misc/proxmox_domain_info.py]: Fixed.

**Comment by Aversiste** [comment]
Rebased and CI OK.

**Comment by felixfontein** [comment]
Please add tests/integration/targets/proxmox/aliases with the following content:
```
unsupported
proxmox_domain_info
proxmox_group_info
proxmox_user_info
```

File: plugins/modules/cloud/misc/proxmox_user_info.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: proxmox_user_info
+short_description: Retrieve information about one or more Proxmox VE users
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_user_info.py:14]:
```suggestion
short_description: Retrieve information about one or more Proxmox VE users
version_added: 1.3.0
```

File: plugins/modules/cloud/misc/proxmox_group_info.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: proxmox_group_info
+short_description: Retrieve information about one or more Proxmox VE groups
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_group_info.py:14]:
```suggestion
short_description: Retrieve information about one or more Proxmox VE groups
version_added: 1.3.0
```

File: plugins/modules/cloud/misc/proxmox_domain_info.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: proxmox_domain_info
+short_description: Retrieve information about one or more Proxmox VE domains
```
Comment (felixfontein) [inline plugins/modules/cloud/misc/proxmox_domain_info.py:14]:
```suggestion
short_description: Retrieve information about one or more Proxmox VE domains
version_added: 1.3.0
```

File: plugins/doc_fragments/proxmox.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    description:
+      - Specify the token secret.
+    type: str
+    version_added: 1.3.0
+  validate_certs:
+    description:
+      - If C(no), SSL certificates will not be validated.
+      - This should only be used on personally controlled sites using self-signed certificates.
+    type: bool
+    default: 'no'
```
Comment (felixfontein) [inline plugins/doc_fragments/proxmox.py:43]:
```suggestion
    default: no
```

**Comment by felixfontein** [comment]
(Failing CI is unrelated.)

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@Aversiste tell me if I should merge this now.

**Comment by Aversiste** [comment]
Yes you can :)

**Comment by felixfontein** [comment]
@Aversiste thanks a lot for working on this!

**Pull request merged**